[documentation] #2193: Update Iroha Client and Iroha CLI Client documentation (#2234)
hyperledger/iroha: client_cli/README.md

# Iroha CLI Client

With Iroha CLI Client, you can interact with Iroha Peers Web API without directly using the network.

> It's a "light" client, meaning it only converts Command Line Interface commands into Iroha Web API Network Requests.

Reviewer: ```suggestion
It is a "thin" wrapper around functionality exposed in the `iroha_client` crate. Specifically, it should be used as a reference for using `iroha_client`'s features, and not as a production-ready client. As such, the CLI client  is not guaranteed to support all features supported by the client library. 
```

## Features

* Submit Transactions with your Iroha Special Instructions to Iroha Peers
* Send Requests with your Queries to Iroha Peers

## Installation

> **Requirements:** a working [Rust toolchain](https://www.rust-lang.org/learn/get-started) installed.

Reviewer: ```suggestion
**Requirements:** a working [Rust toolchain](https://www.rust-lang.org/learn/get-started) (version 1.60) installed and configured.
```

Reviewer: As an alternative.

Build Iroha and its binaries:

```bash
cargo build
```

> It will produce `iroha_client_cli` executable file for Unix and `iroha_client_cli.exe` executable file for Windows.

Reviewer: ```suggestion
The above command will produce the `iroha_client_cli`  ELF executable file for Linux/BSD, the `iroha_client_cli` executable for MacOS,  and the `iroha_client_cli.exe` executable for Windows, depending on your platform and configuration. 
```

## Usage Examples

> :grey_exclamation: All examples below are Unix-oriented. To run them on Windows, replace `./iroha_client_cli` with `iroha_client_cli.exe`.

Reviewer: ```suggestion
:grey_exclamation: All examples below are Unix-oriented. If you're on Windows, we would highly encourage you to consider using WSL, as most documentation assumes a POSIX-like shell running on your system. Please be advised that the syntax of commands may change more than just executing `iroha_client_cli.exe`  instead of `iroha_client_cli`. 
```

In this section we will show you how to use Iroha CLI Client to do the following:

> - [Create new domain](#create-new-domain)

Reviewer: ```suggestion
- [Create a new domain](#create-new-domain)
```

> - [Create new account](#create-new-account)

Reviewer: ```suggestion
- [Create a new account](#create-new-account)
```

> - [Mint asset to account](#mint-asset-to-account)

Reviewer: ```suggestion
- [Mint an asset to an account's wallet](#mint-asset-to-account)
```

> - [Query account assets quantity](#query-account-assets-quantity)

Reviewer: ```suggestion
- [Query an account assets](#query-account-assets-quantity)
```

### Create new Domain

Let's start with domain creation. To create a domain, you need to specify the entity type first (`domain` in our case) and then the command (`register`) with a list of required parameters.

> For domain entity you only need to provide the `id` parameter as a string.

Reviewer: ```suggestion
For the `domain` entity, you only need to provide the `id` argument as a string that doesn't contain either: 
- whitespace
- the "@" symbol.
- the "#" symbol. 
```

```bash
./iroha_client_cli domain register --id="Soramitsu"
```

Now you have a domain without any accounts.

### Create new Account

> Let's create a new account. Like in the previous example, specify the entity type (`account`) and the command (`register`). Then define the account name with the `id` flag.

Reviewer: ```suggestion
Let's create a new account. Like in the previous example, specify the entity type (`account`) and the command (`register`). Then define the account name as the value of the  `id` argument.
```

> You also need to provide the `key` argument with the account's public key as a double-quoted

Reviewer: ```suggestion
Additionally, you need to provide the `key` argument with the account's public key as a double-quoted multihash representation of the key. Providing an empty string also works (but is highly discouraged), while  omitting the argument altogether will produce an error. 
```

string.

```bash
./iroha_client_cli account register --id="White Rabbit@Soramitsu" --key=""
```

### Mint Asset to Account

> It's time to give something to the Account you created. Let's add some Assets quantity to it.

Reviewer: ```suggestion
It's time to give something to the Account you created. Let's add some Assets of the type `Quantity` to the account.
```

> This time, you need to register an Asset Definition first and then add some Assets to the account. To do this, specify the `asset` entity and then use the `register` and `mint` commands respectively. Every asset has its own value type. In this example, it is defined as quantity (integer/number).

Reviewer: ```suggestion
To do so, you must first register an Asset Definition and only then add some Assets to the account. Specify the `asset` entity and then use the `register` and `mint` commands respectively. 

Every asset has its own value type. In this example, it is defined as `Quantity`, a 32-bit unsigned integer. We also support `BigQuantity` and `Fixed` which are a 128 bit unsigned integer and a 64-bit fixed-precision binary fraction.
```

```bash
./iroha_client_cli asset register --id="XOR#Soramitsu" --value-type=Quantity
./iroha_client_cli asset mint --account="White Rabbit@Soramitsu" --asset="XOR#Soramitsu" --quantity=1010 
```

> You created `XOR#Soramitsu` asset of type `Quantity` and then gave `1010` units of this asset to the `White Rabbit@Soramitsu` account.

Reviewer: ```suggestion
You created `XOR#Soramitsu`, an asset of type `Quantity` and then gave `1010` units of this asset to the account `White Rabbit@Soramitsu`. 
```

### Query Account Assets Quantity

> Because distributed systems heavily relay on the concept of eventual consistency and Iroha works in Consensus between peers, your requests may or may not be processed

Reviewer: ```suggestion
Because distributed systems heavily rely on the concept of _eventual_ consistency and Iroha works by awaiting consensus between peers, your request is not guaranteed to be processed (or be accepted) even if it is correctly formed. 
```

> while Iroha Client will successfully send them and Iroha Peer will accept them. Different stages of transaction processing and different cases may lead to

Reviewer: ```suggestion
While the Iroha Client will successfully send your transactions and the  Iroha Peer will confirm receiving them, it is possible that your request will not appear in the next block. Different causes such as a transaction timeout, a faulty peer in the network, catastrophic failure of the peer that you've sent your data towards and many other conditions naturally occurring inside of any blockchain may lead to a rejection of your transaction at many different stages of processing. It should be noted that Iroha is designed to reduce the incidence of such rejections, and only rejects properly formed transactions in situations when not rejecting it would lead to data corruption and a hard-fork of the network. 
```

rejection of a transaction after your receive response from Command Line Interface. 

> To check that your instructions were applied and the system now in the desired state,

Reviewer: ```suggestion
As such it's important to check that your instructions were applied and the _world_ is now in the desired state,
```

For the `domain` entity, you only need to provide the `id` argument as a string that doesn't contain the `@` and `#` symbols.

```bash
./iroha_client_cli domain register --id="Soramitsu"
```

Now you have a domain without any accounts.

### Create new Account

Let's create a new account. Like in the previous example, specify the entity type (`account`) and the command (`register`). Then define the account name as the value of the `id` argument.
Additionally, you need to provide the `key` argument with the account's public key as a double-quoted multihash representation of the key. Providing an empty string also works (but is highly discouraged), while omitting the argument altogether will produce an error.

```bash
./iroha_client_cli account register --id="White Rabbit@Soramitsu" --key=""
```

### Mint Asset to Account

It's time to give something to the Account you created. Let's add some Assets of the type `Quantity` to the account.

To do so, you must first register an Asset Definition and only then add some Assets to the account. Specify the `asset` entity and then use the `register` and `mint` commands respectively. 
	

> Every asset has its own value type. In this example, it is defined as `Quantity`, a 32-bit unsigned integer. We also support `BigQuantity` and `Fixed`, which are a 128-bit unsigned integer and a 64-bit fixed-precision binary fraction.

Reviewer 2: There is also another type called [`Store`](https://github.com/hyperledger/iroha/blob/49407ff210a96c63de41038d9ee68007e5f02098/data_model/src/asset.rs#L189). If I understand it right, it can be used to create *NFT* or some other things I can't find an example of
